src/decision/dMCDA.jl

```diff
@@ -737,87 +721,111 @@ function guided_site_selection(
-Replaces these locations with those in the top ranks if the distance between these is greater.
+    constrain_reef_cluster(reefs::Union{Vector{String},Vector{Float64}}, s_order::Matrix{Union{Float64,Int64}}, rankings::Matrix{Int64},
+    area_to_seed::Float64, available_space::Vector{Float64}, n_site_int::Int64, n_reefs::Int64)
 
 # Arguments
-- `pref_locs` : Original n highest ranked locations selected for seeding or shading.
-- `s_order` : Current order of ranked sites in terms of numerical site ID.
-- `dist` : Matrix of unique distances between sites.
-- `min_dist` : Minimum distance between sites for selected sites.
-- `rankings` : Ranking data
-- `rank_col` : Index of column holding location ranks
+- `reefs` : List of the the reefs each location sits within
+- `s_order` : Ordered set of locations and their aggregate criteria score
+- `rankings` : Current ranks of the set of locations
+- `area_to_seed` : absolute area to be seeded (m²)
+- `available_space` : absolute area available at each location (m²)
+- `n_site_int` : Minimum number of sites to intervene at
+- `n_reefs` : Number of selected locations to allow in the same reef.
 
 # Returns
-New set of selected sites for seeding or shading.
+Tuple :
+    - `pref_locs` : Vector, Indices of preferred intervention locations
+    - `rankings` : Matrix[n_sites ⋅ 3] where columns are site_id, seeding_rank, shading_rank
+        Values of 0 indicate sites that were not considered
 """
-function distance_sorting(
-    pref_locs::AbstractArray{Int64},
+function constrain_reef_cluster(
+    reefs::Union{Vector{String},Vector{Float64}},
     s_order::Matrix{Union{Float64,Int64}},
-    dist::Matrix{Float64},
-    min_dist::Float64,
     rankings::Matrix{Int64},
-    rank_col::Int64,
-)::Tuple{Vector{Union{Float64,Int64}},Matrix{Int64}}
-    # set-up
-    n_sites = length(pref_locs)
-    site_order = s_order[:, 1]
-
-    # Sites to select alternatives from
-    alt_sites = setdiff(site_order, pref_locs)
-
-    # Find all selected sites closer than the min distance
-    pref_dists = findall(dist[pref_locs, pref_locs] .< min_dist)
-
-    idx_to_replace = sort(unique(reinterpret(Int64, pref_dists)))
-    select_n = length(idx_to_replace)
-
-    keep_idxs = setdiff(collect(1:length(pref_locs)), idx_to_replace)
-
-    # Store of new set of locations
-    updated_locs = pref_locs
+    area_to_seed::Float64,
+    available_space::Vector{Float64},
+    n_site_int::Int64,
+    n_reefs::Int64,
+)
+    # Get full ordering of locations
+    loc_order = s_order[:, 1]
+    # Get unique reefs in location set
+    unique_reefs = unique(reefs)
+    pref_locs = loc_order[1:n_site_int]
+
+    num_locs = n_site_int # No. of locations to consider
+
+    for ll in 1:length(loc_order)
+        # If enough space for seeding corals, keep n_site_int, else expand as needed
+        cumulative_space = cumsum(available_space[loc_order])
```

Comment:
good point, `loc_order` changes but the areas will not, I'll move this outside

Reply:
actually this does need to be inside the loop, because `cumsum` is order dependent

Reply:
I don't follow here, as `loc_order` does not change, neither does `available_space`, and is not dependent on `ll`.
So I'm not sure what ordering is changed? Could you elaborate?

Reply:
`loc_order` will change, it's updated when `pref_locs` is updated at each iteration (this will in reality probably only occur once or twice). `loc_order` captures the full order of sites, so only the first 5 site ids or so will change at each iteration, but this will effect the cumulative sum vector. You probably only need to evaluate the first `n_iv_loc*2` entries though to work out how many sites to choose... `ll` is a dummy iterator, it doesn't index anything. I could use `while` instead but I usually avoid using `while`. Let me know if you have a preference.

Reply:
Sorry, I had a look at the full implementation.

This is an issue - you cannot update a variable while looping over it. Only the initial length will be considered.

Even if you could it would introduce other errors.
The only reason it seems to work in your case, I suspect, is that you're only testing situations where the location list can be resolved in a single loop.

```julia
x = rand(10)
for i in 1:length(x)
    x = rand(15)
    @info i
end
```

But specific to what you're trying to do here, I'm not sure it's even necessary.

The suggested change below:

`num_locs = max(findfirst(>=(area_to_seed), cumulative_space), n_site_int)`

Identifies `num_locs` that satisfy requirements already, no?
Meaning you only need to be looping to ensure the reefs are swapped out as necessary.

Reply:
The suggested change works though only if `cumulative_space` is in the same order as the ranking order. If  `cumulative_space` is in the site id order, it doesn't tell you whether the sites you have selected have enough space, only that the first `num_locs` sites have enough space, regardless of their rank. What do you mean by `you cannot update a variable while looping over it'?

Reply:
> The suggested change works though only if cumulative_space is in the same order as the ranking order.

Then, why isn't `cumulative_space` in ranked order? Doesn't it make sense to take the space associated with each desirable location?

> What do you mean by "you cannot update a variable while looping over it"?

You expect the loop as defined by `for ll in 1:length(loc_order)` to be updated because you update `loc_order`.
As shown in my example above, updating `loc_order` will not affect the number of iterations.

Reply:
`cumulative_space` is in ranked order when it is indexed by `loc_order`.`loc_order` will not affect the number of iterations but the loop will be broken out of is `reefs_swap` is empty : https://github.com/open-AIMS/ADRIA.jl/blob/39f20837d33c0067e8eea9af53a9d8b3b2a8ae9a/src/decision/dMCDA.jl#L776
`reefs_swap` will be empty only if the updated selected locations are within the reef constraint.

Reply:
Let's chat about it